---
description: >-
  Guidelines for importing Mothur-generated data into the MicrobiomeStat
  framework.
---

# Fetching Data from Mothur into MicrobiomeStat

This guide details the procedure to seamlessly integrate data generated by Mothur into the MicrobiomeStat environment.

{% file src="../../.gitbook/assets/esophagus.fn.list.gz" %}
This compressed file contains the list file from Mothur, which typically encompasses OTU clustering information for various thresholds.
{% endfile %}

{% file src="../../.gitbook/assets/esophagus.fn.shared.gz" %}
The shared file provides a breakdown of the abundance of OTUs in each sample.
{% endfile %}

{% file src="../../.gitbook/assets/esophagus.good.groups.gz" %}
Group file indicating the sample assignment for each sequence.
{% endfile %}

{% file src="../../.gitbook/assets/esophagus.tree.gz" %}
Phylogenetic tree file representing the evolutionary relationships among OTUs.
{% endfile %}

```r
# Make sure the plyr package is installed
if (!require(plyr)) {
    install.packages("plyr")
    library(plyr)
} else {
    library(plyr)
}

# Specify the paths to your Mothur files
path_to_list_file <- "/path_to_your_files/esophagus.fn.list.gz"
path_to_group_file <- "/path_to_your_files/esophagus.good.groups.gz"
path_to_tree_file <- "/path_to_your_files/esophagus.tree.gz"
path_to_shared_file <- "/path_to_your_files/esophagus.fn.shared.gz"

# Import Mothur data into a MicrobiomeStat data object
data_obj <- mStat_import_mothur_as_data_obj(
    mothur_list_file = path_to_list_file,
    mothur_group_file = path_to_group_file,
    mothur_tree_file = path_to_tree_file,
    mothur_shared_file = path_to_shared_file
)

# After constructing data_obj, detach the plyr package to avoid conflicts with dplyr
if ("package:plyr" %in% search()) {
    detach("package:plyr", unload = TRUE)
    message("Detached plyr package to avoid potential conflicts with dplyr.")
}
```

The `mStat_import_mothur_as_data_obj` function facilitates the conversion of the Mothur datasets to a structured format suitable for MicrobiomeStat. Here are the key components:

* **mothur\_list\_file** (Optional): Mothur-generated list file.
* **mothur\_group\_file** (Optional): Group designation file from Mothur.
* **mothur\_tree\_file** (Optional): Phylogenetic tree representation.
* **mothur\_shared\_file** (Optional): Abundance distribution of OTUs across samples.
* **mothur\_constaxonomy\_file** (Optional): Classification of OTUs at various taxonomic levels.
* **parseFunction** (Optional): Custom function for taxonomy parsing, with 'parse\_taxonomy\_default' as an available option.

Once imported into MicrobiomeStat, the resultant data object can then be employed in diverse downstream analyses facilitated by the MicrobiomeStat toolkit.

<figure><img src="../../.gitbook/assets/Screenshot 2023-10-10 at 16.27.19.png" alt=""><figcaption><p>Structural overview of the `data_obj` generated after importing Mothur datasets into MicrobiomeStat.</p></figcaption></figure>

Commence your analysis in the MicrobiomeStat environment, utilizing its comprehensive functionalities tailored for microbiome research.
